src/cli.ts

```diff
@@ -117,10 +118,58 @@ function debugCliInfo(version: string, channel: string, env: Env, config: IConfi
   );
 }
 
+class VersionObject {
+  public cliVersion: string;
+  public pluginVersions: string[];
+  public osVersion: string;
+
+  public constructor(cliVersion: string, pluginVersions: string[], osVersion: string) {
+    this.cliVersion = cliVersion;
+    this.pluginVersions = pluginVersions;
+    this.osVersion = osVersion;
+  }
+}
+
+class SfdxMain extends Main {
+  // Function which returns Version object which includes cli version, plugin versions, OS and its version.
+  protected getIssueVersionObject(): VersionObject {
+    const cliVersion = this.config.userAgent;
+    const pluginVersion: string = exec('sfdx plugins --core', {
+      silent: true,
+    }).toString();
+    const pluginVersions: string[] = pluginVersion.split('\n');
+    pluginVersions.pop();
+    const osVersion = `${os.type()} ${os.release()}`;
+    return new VersionObject(cliVersion, pluginVersions, osVersion);
+  }
+
+  protected _version(): never {
+    const options: Set<string> = new Set(this.argv);
+
+    // Checking if options has json or verbose
+    if (options.has('--verbose') || options.has('--json')) {
+      const versionObject: VersionObject = this.getIssueVersionObject();
+      if (options.has('--json')) {
+        this.log(`${JSON.stringify(versionObject, null, '\t')}`);
+      } else {
+        this.log(` CLI Version : \n\t${versionObject.cliVersion}`);
+        this.log('\n Plugin Version: ');
+        versionObject.pluginVersions.forEach((plugin) => {
+          this.log(`\t${plugin}`);
+        });
+        this.log(`\n OS and Version: \n\t${versionObject.osVersion}`);
+      }
+    } else {
```

Comment:
just a style preference, but I'd put this first.  ex:

```
if (!options.has('--verbose') && !options.has('--json')) {
  this.log(this.config.userAgent);
  return; // assuming we don't have to use exit codes in oclif context 
}
```

That gets you out early and avoids having nested if/else (or any else at all!)

Reply:
Updated based on the suggestion.